hooks/bootstrap.py

```diff
@@ -0,0 +1,79 @@
+# Copyright (c) 2018 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+from sgtk import get_hook_baseclass
+
+
+class BootstrapHook(get_hook_baseclass()):
+    """
+    The bootstrap hook can be added to a configuration's core hooks folder
+    (``core/hooks``) in order to override some of the functionality of
+    the ToolkitManager's bootstrapping process.
+
+    The hook will be instantiated only after a configuration has been selected
+    by the ``ToolkitManager``. Therefore, this hook will not be invoked to download
+    a configuration. However, the Toolkit Core, applications, frameworks and
+    engines will be downloaded through the hook.
+    """
+
+    def init(self, shotgun, pipeline_configuration_id, configuration_descriptor, **kwargs):
+        """
+        This method is invoked when the hook is instantiated and provides some information
+        about the site and the configuration that are being bootstrapped into.
+
+        The ``kwargs`` is there for forward compatibility with future versions of Toolkit.
+
+        :param shotgun: Connection to the Shotgun site.
+        :type shotgun: :class:`shotgun_api3.Shotgun`
+
+        :param pipeline_configuration_id: Id of the pipeline configuration we're bootstrapping into.
+            If None, the ToolkitManager is bootstrapping into the base configuration.
+        :type pipeline_configuration_id: int
+
+        :param configuration_descriptor: Configuration the manager is bootstrapping into.
+        :type configuration_descriptor: :class:`sgtk.descriptor.ConfigDescriptor`
+        """
+        self.shotgun = shotgun
```

Comment:
do we need this state? I reckon we should delete it.

Reply:
The tk instance or the sgtk.get_authenticated_user is not ready yet when we are caching core, so we need to pass in a shotgun connection. We can't so `self.parent.shotgun`.